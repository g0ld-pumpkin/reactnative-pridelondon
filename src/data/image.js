// @flow
import * as decode from "../lib/decode";
import type { Decoder } from "../lib/decode";

export type Images = {
  [id: string]: ImageDetails
};

export type ImageDetails = {
  id: string,
  revision: number,
  uri: string,
  width: number,
  height: number
};

export const getImageDetails = (images: Images) => (
  id: string,
  dimensions: ?{ width: number, height: number }
): ?ImageDetails => {
  const imageDetails = images[id];

  // Image details might be undefined in case of unpublished image
  if (!dimensions || !imageDetails) {
    return imageDetails;
  }

  const imageUriWithResizingBehaviourParameters = `${imageDetails.uri}?w=${
    dimensions.width
  }&h=${dimensions.height}&fit=fill`;
  return Object.assign({}, imageDetails, {
    uri: imageUriWithResizingBehaviourParameters
  });
};

export const decodeImageDetails = (locale: string): Decoder<ImageDetails> =>
  decode.shape({
    id: decode.at(["sys", "id"], decode.string),
    revision: decode.at(["sys", "revision"], decode.number),
    uri: decode.map(
      value => `https:${value}`,
      decode.at(["fields", "file", locale, "url"], decode.string)
    ),
    height: decode.at(
      ["fields", "file", locale, "details", "image", "height"],
      decode.number
    ),
    width: decode.at(
      ["fields", "file", locale, "details", "image", "width"],
      decode.number
    )
  });
