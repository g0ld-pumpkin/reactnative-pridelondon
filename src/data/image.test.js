// @flow
import type { ImageDetails } from "./image";
import {
  generateImageDetails,
  generateCMSImage,
  sampleOne
} from "./__test-data";
import { decodeImageDetails, getImageDetails } from "./image";

describe("image", () => {
  describe("getImageDetails", () => {
    it("returns the correct imageDetails", () => {
      const imageDetails: ImageDetails = sampleOne(generateImageDetails);
      const getter = getImageDetails({
        [imageDetails.id]: imageDetails
      });

      expect(getter(imageDetails.id)).toEqual(imageDetails);
    });

    it("returns the correct imageDetails when dimensions are provided", () => {
      const imageDetails: ImageDetails = sampleOne(generateImageDetails);
      const imageDimensions = { width: 500, height: 200 };
      const imageDetailsWithDimensions = {
        ...imageDetails,
        uri: `${imageDetails.uri}?w=${imageDimensions.width}&h=${
          imageDimensions.height
        }&fit=fill`
      };

      const getter = getImageDetails({
        [imageDetails.id]: imageDetails
      });

      expect(getter(imageDetails.id, imageDimensions)).toEqual(
        imageDetailsWithDimensions
      );
    });
  });

  it("returns correctly when dimensions are provided but there is no image details for image id", () => {
    const imageDimensions = { width: 500, height: 200 };
    const imageId = "image-id";

    const getter = getImageDetails({});

    expect(getter(imageId, imageDimensions)).toBeUndefined();
  });

  describe("decodeImageDetails", () => {
    it("correctly decodes valid CMS image", () => {
      const data: mixed = sampleOne(generateCMSImage);

      const decoded = decodeImageDetails("en-GB")(data);
      expect(decoded.ok).toEqual(true);
      if (decoded.ok) {
        expect(decoded.value).toMatchSnapshot();
      }
    });

    it("fails if a property is missing", () => {
      const data: mixed = {
        fields: {},
        sys: {
          id: "3O3SZPgYl2MUEWu2MoK2oi"
        }
      };

      const decoded = decodeImageDetails("en-GB")(data);
      expect(decoded.ok).toEqual(false);
    });
  });
});
