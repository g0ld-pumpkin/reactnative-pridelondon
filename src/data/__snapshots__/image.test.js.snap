// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`image decodeImageDetails correctly decodes valid CMS image 1`] = `
Object {
  "height": 985,
  "id": "hG",
  "revision": 1,
  "uri": "https://red-badger.com/BhabSLtI692l05Qliy85.jpg",
  "width": 890,
}
`;
